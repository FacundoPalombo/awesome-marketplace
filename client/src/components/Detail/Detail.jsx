import PropTypes from "prop-types";
import React from "react";
import styles from "./Detail.module";

export function Detail({
  id,
  title,
  price,
  condition,
  soldQuantity,
  categories,
  description,
  thumbnail,
}) {
  return (
    <div className={styles.container}>
      <div className={styles.content}>
        <div className={styles.imageContainer}>
          <img src={thumbnail} className={styles.thumbnail} />
        </div>
        <h4 className={styles.descriptionTitle}>Descripci√≥n del producto</h4>
        <p className={styles.description}>{description}</p>
      </div>
      <div className={styles.heading}>
        <p className={styles.condition}>
          {condition === "new" ? "Nuevo" : "Usado"} - {soldQuantity} vendidos
        </p>

        <h4 className={styles.title}>{title}</h4>
        <h3 className={styles.price}>{price}</h3>
        <button className={styles.cta}>Comprar</button>
      </div>
    </div>
  );
}

Detail.propTypes = {
  id: PropTypes.string,
  title: PropTypes.string,
  price: PropTypes.string,
  condition: PropTypes.string,
  sold_quantity: PropTypes.number,
  categories: PropTypes.arrayOf(
    PropTypes.shape({ name: PropTypes.string, id: PropTypes.string })
  ),
  description: PropTypes.string,
  thumbnail: PropTypes.string,
};
