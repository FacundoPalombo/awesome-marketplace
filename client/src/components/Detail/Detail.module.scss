@import "utils/variables.scss";
.container {
  display: flex;
  flex-direction: row;

  @include xl {
    flex-direction: column;
  }
}

.content {
  display: flex;
  flex-direction: column;
  width: 65%;
  align-items: space-around;
  padding: $unit-4 0 $unit-4 $unit-4;
  @include xl {
    width: 80%;
  }
}

.imageContainer {
  width: 100%;
  height: auto;
  box-sizing: content-box;
}

.thumbnail {
  width: 100%;
  max-width: 680px;
  max-height: 680px;
  @include xl {
    box-sizing: content-box;
  }
}

.price {
  margin-block: $unit-4;
  font-size: 46px;
  font-weight: 400;
  &::before {
    content: "$";
    padding-right: $unit;
  }
}

.heading {
  display: flex;
  flex-direction: column;
  padding: $unit-4 $unit-4 $unit-4 $unit-4;
  width: 30%;
  align-items: flex-start;
  justify-content: flex-start;
  @include xl {
    width: 100%;
    padding: $unit-4;
    box-sizing: border-box;
  }
}

.title {
  width: 100%;
  margin-block: 0;
  font-weight: 400;
  font-size: 24px;
}

.condition {
  font-size: 14px;
  margin-block: 0 $unit-2;
}

.cta {
  padding: $unit;
  background-color: $cta;
  outline: none;
  border: none;
  border-radius: $half-unit;
  color: #ffffff;
  width: 100%;
  @include xl {
    align-self: center;
    width: 80%;
  }
}

.description {
  font-size: 16px;
  text-overflow: ellipsis;
  &Title {
    font-weight: 500;
    font-size: 28px;
  }
}
