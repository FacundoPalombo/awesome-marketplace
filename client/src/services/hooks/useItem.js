import axios from "axios";
import { useEffect, useState } from "react";

export function useItem(id) {
  const [item, setItem] = useState({});
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);
  useEffect(() => {
    async function getItem(id) {
      setItem({});
      setError(null);
      setLoading(true);
      await axios
        .get(`http://0.0.0.0:3001/api/item/${id}`)
        .then((response) => {
          let priceFormatted = response.data.item.price.toLocaleString("es-AR");
          setLoading(false);
          setItem({ ...response.data.item, ...{ price: priceFormatted } });
        })
        .catch((err) => {
          setLoading(false);
          setError({
            code: err.response.status,
            message: err.response.statusText,
          });
        });
    }
    getItem(id);
  }, [id]);
  return [item, loading, error];
}
