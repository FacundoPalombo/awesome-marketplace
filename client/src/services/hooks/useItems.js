import axios from "axios";
import { useEffect, useState } from "react";

export function useItems(query) {
  const [items, setItems] = useState({});
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);
  useEffect(() => {
    async function getItems(query) {
      setLoading(true);
      const response = await axios.get(
        `http://0.0.0.0:3001/api/items?q=${query}`
      );
      setLoading(false);
      setItems(response.data);
      if (response.status >= 400) {
        setLoading(false);
        setError({ code: response.status, message: response.statusText });
      }
    }
    getItems(query);
  }, [query]);
  return [items, loading, error];
}
