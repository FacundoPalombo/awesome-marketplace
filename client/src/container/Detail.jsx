import React, { useEffect } from "react";
import { useParams } from "react-router-dom";
import { useItem } from "services/hooks/useItem";
import styles from "./Detail.module";
import { Breadcrumbs } from "common/Breadcrumbs/Breadcrumbs";
import { Detail as DetailWrapper } from "components/Detail/Detail";
import { Helmet } from "react-helmet";
import Spinner from "react-spinners/BounceLoader";

export function Detail() {
  const { id } = useParams();
  const [item, loading, error] = useItem(id);

  return (
    <section className={styles.container}>
      <div className={styles.breadcrumbsContainer}>
        <nav className={styles.breadcrumbs}>
          {item?.categories?.length !== 0 && !error && (
            <Breadcrumbs categories={item.categories} />
          )}
          {/* TODO: improve this xD */}
        </nav>
      </div>
      {loading && <Spinner size="100" color="#666" />}
      {error && (
        <pre>
          Error {error.code} - {error.message}
        </pre>
      )}
      {Object.entries(item).length !== 0 ? (
        <div className={styles.contentContainer}>
          <Helmet>
            <meta http-equiv="Content-Type" content="text/html;charset=UTF-8" />
            <title>Detalle del producto</title>
          </Helmet>
          <main className={styles.content} id="detail">
            <DetailWrapper
              id={item.id}
              title={item.title}
              price={item.price}
              condition={item.condition}
              soldQuantity={item.sold_quantity}
              categories={item.categories}
              description={item.description}
              thumbnail={item.thumbnail}
            />
          </main>
        </div>
      ) : null}
    </section>
  );
}
