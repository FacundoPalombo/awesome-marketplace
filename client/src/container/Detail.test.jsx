import React from "react";
import { Detail } from "./Detail";
import mockItem from "utils/tests/fixtures/item";
import { useItem } from "services/hooks/useItem";
import { renderWithRouter } from "../utils/tests/renderWithRouter";

jest.mock("services/hooks/useItem", () => ({
  useItem: jest.fn(),
}));

describe("GIVEN Detail Component", () => {
  describe("WHEN page is rendered", () => {
    it("THEN should match to snapshot", () => {
      useItem.mockReturnValue([mockItem, false, null]);
      const { asFragment } = renderWithRouter(<Detail />);
      expect(asFragment()).toMatchSnapshot();
    });
    it("THEN should render the loader", () => {
      useItem.mockReturnValue([{ categories: [] }, true, null]);
      const { asFragment } = renderWithRouter(<Detail />);
      expect(asFragment()).toMatchSnapshot();
    });
    it("THEN should render the empty state", () => {
      useItem.mockReturnValue([{}, false, null]);
      const { asFragment } = renderWithRouter(<Detail />);
      expect(asFragment()).toMatchSnapshot();
    });
    it("THEN should render the error", () => {
      useItem.mockReturnValue([
        { categories: [] },
        false,
        { message: "Oops. Snoop error", code: "420" },
      ]);
      const { asFragment } = renderWithRouter(<Detail />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
