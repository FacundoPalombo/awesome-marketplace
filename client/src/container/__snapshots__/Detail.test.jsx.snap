// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN Detail Component WHEN page is rendered THEN should match to snapshot 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <div
      class="breadcrumbsContainer"
    >
      <nav
        class="breadcrumbs"
      >
        <nav
          class="breadcrumbs"
        />
      </nav>
    </div>
    <div
      class="contentContainer"
    >
      <main
        class="content"
        id="detail"
      >
        <div
          class="container"
        >
          <div
            class="content"
          >
            <img
              class="thumbnail"
            />
            <h4
              class="descriptionTitle"
            >
              Descripción del producto
            </h4>
            <p
              class="description"
            />
          </div>
          <div
            class="heading"
          >
            <p
              class="condition"
            >
              Usado -  vendidos
            </p>
            <h4
              class="title"
            />
            <h3
              class="price"
            />
            <button
              class="cta"
            >
              Comprar
            </button>
          </div>
        </div>
      </main>
    </div>
  </section>
</DocumentFragment>
`;

exports[`GIVEN Detail Component WHEN page is rendered THEN should render the empty state 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <div
      class="breadcrumbsContainer"
    >
      <nav
        class="breadcrumbs"
      >
        <nav
          class="breadcrumbs"
        />
      </nav>
    </div>
  </section>
</DocumentFragment>
`;

exports[`GIVEN Detail Component WHEN page is rendered THEN should render the error 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <div
      class="breadcrumbsContainer"
    >
      <nav
        class="breadcrumbs"
      />
    </div>
    <pre>
      Error 420 - Oops. Snoop error
    </pre>
    <div
      class="contentContainer"
    >
      <main
        class="content"
        id="detail"
      >
        <div
          class="container"
        >
          <div
            class="content"
          >
            <img
              class="thumbnail"
            />
            <h4
              class="descriptionTitle"
            >
              Descripción del producto
            </h4>
            <p
              class="description"
            />
          </div>
          <div
            class="heading"
          >
            <p
              class="condition"
            >
              Usado -  vendidos
            </p>
            <h4
              class="title"
            />
            <h3
              class="price"
            />
            <button
              class="cta"
            >
              Comprar
            </button>
          </div>
        </div>
      </main>
    </div>
  </section>
</DocumentFragment>
`;

exports[`GIVEN Detail Component WHEN page is rendered THEN should render the loader 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <div
      class="breadcrumbsContainer"
    >
      <nav
        class="breadcrumbs"
      />
    </div>
    <span
      class="css-m2td1u"
    >
      <span
        class="css-132xq7w"
      />
      <span
        class="css-16x3gmc"
      />
    </span>
    <div
      class="contentContainer"
    >
      <main
        class="content"
        id="detail"
      >
        <div
          class="container"
        >
          <div
            class="content"
          >
            <img
              class="thumbnail"
            />
            <h4
              class="descriptionTitle"
            >
              Descripción del producto
            </h4>
            <p
              class="description"
            />
          </div>
          <div
            class="heading"
          >
            <p
              class="condition"
            >
              Usado -  vendidos
            </p>
            <h4
              class="title"
            />
            <h3
              class="price"
            />
            <button
              class="cta"
            >
              Comprar
            </button>
          </div>
        </div>
      </main>
    </div>
  </section>
</DocumentFragment>
`;
